### ISSUE TITLE
[Feature Request] Missing data likelihoods

### ISSUE THREAD

**Issue by InfProbSciX** [cornellius-gp/gpytorch#1591]
# 🚀 Feature Request

We'd like to use GPs in settings where some observations may be missing. My understanding is that, in these circumstances, missing observations do not contribute anything to the likelihood of the observation model.

## Initial Attempt

My initial attempt to write such a likelihood is as follows:

```python
from gpytorch.likelihoods import GaussianLikelihood
from torch.distributions import Normal

class GaussianLikelihoodWithMissingObs(GaussianLikelihood):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    @staticmethod
    def _get_masked_obs(x):
        missing_idx = x.isnan()
        x_masked = x.masked_fill(missing_idx, -999.)
        return missing_idx, x_masked

    def expected_log_prob(self, target, input, *params, **kwargs):
        missing_idx, target = self._get_masked_obs(target)
        res = super().expected_log_prob(target, input, *params, **kwargs)
        return res * ~missing_idx

    def log_marginal(self, observations, function_dist, *params, **kwargs):
        missing_idx, observations = self._get_masked_obs(observations)
        res = super().log_marginal(observations, function_dist, *params, **kwargs)
        return res * ~missing_idx
```

## Test

```python
import torch
import numpy as np
from tqdm import trange
from gpytorch.distributions import MultivariateNormal
from gpytorch.constraints import Interval
torch.manual_seed(42)

mu = torch.zeros(2, 3)
sigma = torch.tensor([[
        [ 1,  1-1e-7, -1+1e-7],
        [ 1-1e-7,  1, -1+1e-7],
        [-1+1e-7, -1+1e-7,  1] ]]*2).float()
mvn = MultivariateNormal(mu, sigma)
x = mvn.sample_n(10000)
# x[np.random.binomial(1, 0.1, size=x.shape).astype(bool)] = np.nan
x += np.random.normal(0, 0.5, size=x.shape)

LikelihoodOfChoice = GaussianLikelihood#WithMissingObs
likelihood = LikelihoodOfChoice(noise_constraint=Interval(1e-6, 2))

opt = torch.optim.Adam(likelihood.parameters(), lr=0.5)

bar = trange(1000)
for _ in bar:
    opt.zero_grad()
    loss = -likelihood.log_marginal(x, mvn).sum()
    loss.backward()
    opt.step()
    bar.set_description("nll: " + str(int(loss.data)))
print(likelihood.noise.sqrt()) # Test 1

likelihood.expected_log_prob(x[0], mvn) == likelihood.log_marginal(x[0], mvn) # Test 2
```

`Test 1` outputs the correct `0.5` as expected, and `Test 2` is `False` with `LikelihoodOfChoice = GaussianLikelihood` and `LikelihoodOfChoice = GaussianLikelihoodWithMissingObs`.

Any further tests and suggestions are appreciated. Can I open a PR for this?

**Comment by jacobrgardner** [comment]
Hi @InfProbSciX -- isn't this equivalent to just discarding data points with missing labels ahead of time?

**Comment by InfProbSciX** [comment]
Hi @jacobrgardner, yes it is. For a GPLVM implementation that @vr308 is working on, data is typically fed in as a matrix of dimension (n_data, n_dims) - and this class supports the case where any of these values can be missing.

I believe that discarding the points with missing "y"/output values would be akin to vectorizing this matrix and getting rid of the observations corresponding to missing values, which, as I understand, would be difficult to do / isn't how the GPLVM is set up.

**Comment by vr308** [comment]
Hi @jacobrgardner so this type of likelihood is mainly used in the unsupervised setting (GPLVM based models, I've been working with @gpleiss on implementing it in gpytorch and just submitted a PR). If the high dimensional data Y is modelled as a factorised (over n and d) Gaussian, then one can just marginalise out the missing observations (akin to dropping them). This class supports that operation so one can directly define a likelihood on such data. 

The idea really comes from Section 2.1 of this [paper](https://www.cs.toronto.edu/~urtasun/publications/lawrence_urtasun_icml09.pdf).

**Comment by gpleiss** [comment]
Hi @InfProbSciX and @vr308 - this implementation looks reasonable to me, but I think it would potentially make the most sense as an extension of MultitaskGaussianLikelihood. We're open to a PR!

**Comment by InfProbSciX** [comment]
Sounds good - I believe that the extension to `MultitaskGaussianLikelihood` might look slightly different, but I think I have an idea on how to make it work (by replacing the appropriate rows in the covariance matrices by a diagonal matrix, with the diagonal element being 1/sqrt(2*pi) and filling in the missing elements to zero, so that they have a zero contribution to the log likelihood).

**Comment by InfProbSciX** [comment]
Hi @gpleiss - a quick question, what is expected as the `function_dist` argument of `MultitaskGaussianLikelihood `?

This code throws an error with the `MultitaskGaussianLikelihood `:

```python
import torch
from gpytorch.distributions import MultivariateNormal
from gpytorch.likelihoods import GaussianLikelihood, MultitaskGaussianLikelihood

mu = torch.zeros(2, 3)
sigma = torch.tensor([[
        [ 1,  1-1e-7, -1+1e-7],
        [ 1-1e-7,  1, -1+1e-7],
        [-1+1e-7, -1+1e-7,  1] ]]*2).float()

mvn = MultivariateNormal(mu, sigma)
x = mvn.sample_n(10000)

likelihood = GaussianLikelihood()

likelihood.log_marginal(x, mvn)

likelihood = MultitaskGaussianLikelihood(
    num_tasks=mu.shape[1], rank=mu.shape[1])

likelihood.log_marginal(x, mvn) # error
```

Is this usage incorrect?

**Comment by gpleiss** [comment]
Closed by #1668

**Comment by mochar** [comment]
Hi @InfProbSciX did you ever ended up implementing this as a multi-task gaussian likelihood? I would be appreciative of the code if you have it.

**Comment by InfProbSciX** [comment]
Hey @mochar - unfortunately no, but I'd love to help out. I got as far as that minimal example [above](https://github.com/cornellius-gp/gpytorch/issues/1591#issuecomment-833804967). Do you have any ideas on what the correct usage of the multitask likelihood looks like? If we could get an example such as the above to work, implementing it should be easy, hopefully.

**Comment by mochar** [comment]
> Hey @mochar - unfortunately no, but I'd love to help out. I got as far as that minimal example [above](https://github.com/cornellius-gp/gpytorch/issues/1591#issuecomment-833804967). Do you have any ideas on what the correct usage of the multitask likelihood looks like? If we could get an example such as the above to work, implementing it should be easy, hopefully.

My use case is that for a percentage of the datapoints I only have measurements of a specific subset of the tasks. I am also using a GPLVM model as you are, though I'm not following exactly the model as described in the documentation. Instead I use `LMCVariationalStrategy` as I have hundreds of tasks so this helps speed things up a bit. I also recall not knowing how to train seperate kernel parameters using the GPLVM model as described in the documentation, but I'm digressing..  

Currently I got the model to "work" with missing data by extending `VariationalELBO` and `MultitaskGaussianLikelihood` as follows:

```python
import math

import torch
from gpytorch.lazy import ConstantDiagLazyTensor
from gpytorch.mlls._approximate_mll import _ApproximateMarginalLogLikelihood
from gpytorch.distributions import MultitaskMultivariateNormal
from gpytorch.likelihoods import MultitaskGaussianLikelihood

class HalfMarginalVariationalELBO(_ApproximateMarginalLogLikelihood):
    def __init__(self, likelihood, marginal_indices, model, num_data, **kwargs):
        super().__init__(likelihood, model, num_data, **kwargs)
        self.marginal_indices = torch.LongTensor(marginal_indices)
        # Flag: marginal is not really a marginal, i.e. has all features
        self.marginal_is_subset = self.marginal_indices.size(0) < likelihood.num_tasks

    def _log_likelihood_term(self, variational_dist_f, target, partials, **kwargs):
        n = target.size(-2)
        n_partial = partials.sum()

        if n_partial == 0 or not self.marginal_is_subset:
            return self.likelihood.expected_log_prob(target, variational_dist_f, **kwargs).sum(-1)

        d = target.size(-1)
        idx_all = torch.arange(n * d).reshape(d, n).T
        
        if n_partial < n:
            idx_full = idx_all[~partials, :].T.flatten()
            covar = variational_dist_f.lazy_covariance_matrix[idx_full][:, idx_full]
            mean = variational_dist_f.mean[~partials]
            dist = MultitaskMultivariateNormal(mean, covar)
            a = self.likelihood.expected_log_prob(target[~partials], dist, **kwargs).sum(-1)
        else:
            a = 0

        idx_partial = idx_all[partials][:, self.marginal_indices].T.flatten()
        covar = variational_dist_f.lazy_covariance_matrix[idx_partial][:, idx_partial]
        mean = variational_dist_f.mean[partials][:, self.marginal_indices]
        dist_marg = MultitaskMultivariateNormal(mean, covar)
        b = self.likelihood.expected_log_prob(target[partials][:, self.marginal_indices], dist_marg, **kwargs).sum(-1)

        return a + b

    def forward(self, variational_dist_f, target, **kwargs):
        return super().forward(variational_dist_f, target, **kwargs)
```
Here I copied the implementation of `VariationalELBO` but extended the `_log_likelihood_term` method to accept a boolean tensor `partials` indicating which datapoints have the missing values. For those datapoints I take the marginal distribution of the `MultitaskMultivariateNormal` that my model class returns. With marginal distribution I mean excluding the dimensions that correspond to the tasks that are not measured in the subset of points with missing values. The indices of these tasks, `marginal_indices`, are passed to `__init__`.

However this does not work because `MultitaskGaussianLikelihood.expected_log_prob()` calls the method `_shaped_noise_covar` to get the noise parameters, but this method is not aware that I passed a marginal distribution, so I get a mismatch in shapes (it returns the noise parameters of all tasks, but `target` and `input` both have a subset of the tasks). So I extended this class and commented out the call to `_shaped_noise_covar`, and instead create the Tensor with the noise covar myself. However this only works because I set `has_task_noise=False` and `rank=0`. Otherwise you'd have to select the relevant code from `_shaped_noise_covar`.

```python    
class HalfMarginalMultitaskGaussianLikelihood(MultitaskGaussianLikelihood):
    def expected_log_prob(self, target, input, *params, **kwargs):
        mean, variance = input.mean, input.variance
        num_event_dim = len(input.event_shape)

        # noise = self._shaped_noise_covar(mean.shape, *params, **kwargs).diag()
        noise = ConstantDiagLazyTensor(self.noise, diag_shape=mean.shape[-2] * mean.shape[-1]).diag()

        # Potentially reshape the noise to deal with the multitask case
        noise = noise.view(*noise.shape[:-1], *input.event_shape)

        res = ((target - mean) ** 2 + variance) / noise + noise.log() + math.log(2 * math.pi)
        res = res.mul(-0.5)
        if num_event_dim > 1:  # Do appropriate summation for multitask Gaussian likelihoods
            res = res.sum(list(range(-1, -num_event_dim, -1)))
        return res
```

As you can see this is all extremely hacky (not even sure if it works..) so some simplified way of doing this would be great.

**Comment by mochar** [comment]
@InfProbSciX Any thoughts?

**Comment by InfProbSciX** [comment]
@mochar My apologies, I haven't been receiving notifications - I'll have a look and get back to you.

**Comment by InfProbSciX** [comment]
Hi @mochar - I think I see what's going on here, but could you also please provide a small example of the usage of `HalfMarginalVariationalELBO` and `HalfMarginalMultitaskGaussianLikelihood` (even if it fails, I would like to enter it using a debugger and see what `covar`, `target`, `input` and some other variables look like).

**Comment by mochar** [comment]
Hi @InfProbSciX please have a look here: https://gist.github.com/mochar/3d31726ebf45716746eb33bf96ba5492

**Comment by InfProbSciX** [comment]
Thanks a lot for that! Using your example, I've constructed a minimal example for the likelihood, as I did with the other one:

```python
import torch
import numpy as np
from tqdm import trange
from gpytorch.likelihoods import MultitaskGaussianLikelihood
from gpytorch.distributions import MultitaskMultivariateNormal
torch.manual_seed(42)

mu = torch.zeros(2, 3)
sigma = torch.tensor([[
        [ 1,  1-1e-7, -1+1e-7],
        [ 1-1e-7,  1, -1+1e-7],
        [-1+1e-7, -1+1e-7,  1] ]]*2).float()

likelihood = MultitaskGaussianLikelihood(
    num_tasks=mu.shape[0],
    rank=0,
    has_global_noise=False)

mvn = MultitaskMultivariateNormal(
    mean=mu.T,
    covariance_matrix=torch.kron(torch.eye(2), sigma[0]), interleaved=False)

x = mvn.sample_n(10000)
x += np.random.normal(0, 0.5, size=x.shape)

opt = torch.optim.Adam(likelihood.parameters(), lr=0.5)

bar = trange(250)
for _ in bar:
    opt.zero_grad()
    loss = -likelihood.log_marginal(x, mvn).sum()
    loss.backward()
    opt.step()
    bar.set_description("nll: " + str(int(loss.data)))
print(likelihood.task_noises.sqrt())
```

The issue with this likelihood is that, it seems as though `likelihood.expected_log_prob(x, mvn)` sums the marginals over tasks, so if we mask the observations by some value like `-999` like in the earlier case, we run into issues. I'll have a think about what to do here.

**Comment by mochar** [comment]
Hi @InfProbSciX, I followed [this thread](https://github.com/cornellius-gp/gpytorch/issues/1366) and managed to get it to work I think. It comes down to multiplying the log likelihood values with a mask matrix in the `Likelihood` class. See my notebook [here](https://github.com/mochar/zusenzo/blob/main/gpytorch/masked.ipynb).

**Comment by adamjstewart** [comment]
I am also interested in a `MultitaskGaussianLikelihoodWithMissingObs` for the same reasons listed above (not all data points contain observations for all tasks). Will try to read through the discussion above to get up to speed. @mochar you mentioned that you may have gotten this working, did you open a PR with your code?

One implementation detail I'm not sure about: in the case of each task having a different percentage of missing observations, should we normalize the loss by dividing each task by the total number of observations? It seems like each task should equally contribute to the overall loss.
